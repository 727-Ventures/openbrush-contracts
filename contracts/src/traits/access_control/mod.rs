// Copyright (c) 2012-2022 Supercolony. All Rights Reserved.
// Copyright (c) 2023 Brushfam. All Rights Reserved.
// SPDX-License-Identifier: MIT

#[allow(clippy::module_inception)]
mod access_control;

pub use access_control::*;

pub mod extensions {
    pub mod enumerable;
}
