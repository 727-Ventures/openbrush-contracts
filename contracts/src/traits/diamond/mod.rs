// Copyright (c) 2012-2022 Supercolony. All Rights Reserved.
// Copyright (c) 2023 Brushfam. All Rights Reserved.
// SPDX-License-Identifier: MIT

#[allow(clippy::module_inception)]
mod diamond;

pub use diamond::*;
pub mod extensions {
    pub mod diamond_loupe;
}
