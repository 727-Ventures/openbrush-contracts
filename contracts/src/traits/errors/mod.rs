// Copyright (c) 2012-2022 Supercolony. All Rights Reserved.
// Copyright (c) 2023 Brushfam. All Rights Reserved.
// SPDX-License-Identifier: MIT

mod access_control;
mod diamond;
mod flashloan;
mod ownable;
mod pausable;
mod payment_splitter;
mod psp22;
mod psp34;
mod psp37;
mod reentrancy_guard;
mod timelock_controller;
mod upgradeable;

pub use access_control::AccessControlError;
pub use diamond::DiamondError;
pub use flashloan::{
    FlashBorrowerError,
    FlashLenderError,
};
pub use ownable::OwnableError;
pub use pausable::PausableError;
pub use payment_splitter::PaymentSplitterError;
pub use psp22::{
    PSP22Error,
    PSP22ReceiverError,
    PSP22TokenTimelockError,
};
pub use psp34::{
    PSP34Error,
    PSP34ReceiverError,
};
pub use psp37::{
    PSP37Error,
    PSP37ReceiverError,
};
pub use reentrancy_guard::ReentrancyGuardError;
pub use timelock_controller::TimelockControllerError;
pub use upgradeable::UpgradeableError;
