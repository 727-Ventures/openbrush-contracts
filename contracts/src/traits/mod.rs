// Copyright (c) 2012-2022 Supercolony. All Rights Reserved.
// Copyright (c) 2023 Brushfam. All Rights Reserved.
// SPDX-License-Identifier: MIT

pub mod access_control;
pub mod diamond;
pub mod errors;
pub mod flashloan;
pub mod governance;
pub mod nonces;
pub mod ownable;
pub mod pausable;
pub mod payment_splitter;
pub mod proxy;
pub mod psp22;
pub mod psp34;
pub mod psp37;
pub mod psp61;
pub mod upgradeable;

pub mod types;
