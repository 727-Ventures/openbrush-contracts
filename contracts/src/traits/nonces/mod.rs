// Copyright (c) 2012-2022 Supercolony. All Rights Reserved.
// Copyright (c) 2023 Brushfam. All Rights Reserved.
// SPDX-License-Identifier: MIT

pub use crate::traits::errors::NoncesError;
use openbrush::traits::AccountId;

/// Provides tracking nonces for addresses. Nonces will only increment.
#[openbrush::trait_definition]
pub trait Nonces {
    /// Returns the nonce of `account`.
    #[ink(message)]
    fn nonces(&self, account: AccountId) -> u64;
}
