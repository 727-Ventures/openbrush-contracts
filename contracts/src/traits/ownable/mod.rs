// Copyright (c) 2012-2022 Supercolony. All Rights Reserved.
// Copyright (c) 2023 Brushfam. All Rights Reserved.
// SPDX-License-Identifier: MIT

pub use crate::traits::errors::OwnableError;
use openbrush::traits::AccountId;

#[openbrush::wrapper]
pub type OwnableRef = dyn Ownable;

/// Contract module which provides a basic access control mechanism, where
/// there is an account (an owner) that can be granted exclusive access to
/// specific functions.
#[openbrush::trait_definition]
pub trait Ownable {
    /// Returns the address of the current owner.
    #[ink(message)]
    fn owner(&self) -> Option<AccountId>;

    /// Leaves the contract without owner. It will not be possible to call
    /// owner's functions anymore. Can only be called by the current owner.
    ///
    /// NOTE: Renouncing ownership will leave the contract without an owner,
    /// thereby removing any functionality that is only available to the owner.
    ///
    /// On success a `OwnershipTransferred` event is emitted.
    ///
    /// # Errors
    ///
    /// Panics with `CallerIsNotOwner` error if caller is not owner
    #[ink(message)]
    fn renounce_ownership(&mut self) -> Result<(), OwnableError>;

    /// Transfers ownership of the contract to a `new_owner`.
    /// Can only be called by the current owner.
    ///
    /// On success a `OwnershipTransferred` event is emitted.
    ///
    /// # Errors
    ///
    /// Panics with `CallerIsNotOwner` error if caller is not owner.
    ///
    /// Panics with `NewOwnerIsZero` error if new owner's address is zero.
    #[ink(message)]
    fn transfer_ownership(&mut self, new_owner: Option<AccountId>) -> Result<(), OwnableError>;
}
