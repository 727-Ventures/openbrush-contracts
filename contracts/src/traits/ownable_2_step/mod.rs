pub use crate::traits::errors::OwnableError;
use openbrush::traits::AccountId;

#[openbrush::wrapper]
pub type Ownable2StepRef = dyn Ownable2Step;

/// Contract module which provides a 2 step ownership mechanism, where an
/// owner can transfer ownership to a new address which must then accept
/// the transfer before becoming the new owner.
#[openbrush::trait_definition]
pub trait Ownable2Step {
    #[ink(message)]
    fn pending_owner(&self) -> Option<AccountId>;
    /// Transfers ownership of the contract to a `pending_owner`.
    /// Can only be called by the current owner.
    ///
    /// On success a `OwnershipTransferStarted` event is emitted.
    ///
    /// # Errors
    ///
    /// Panics with `CallerIsNotOwner` error if caller is not owner.
    #[ink(message)]
    fn transfer_ownership(&mut self, new_owner: AccountId) -> Result<(), OwnableError>;

    /// Accepts ownership of the contract.
    /// Can only be called by pending owner.
    ///
    ///
    /// On success a `OwnershipTransferred` event is emitted.
    ///
    /// # Errors
    ///
    /// Panics with `CallerIsNotPendingOwner` error if caller is not pending owner.
    #[ink(message)]
    fn accept_ownership(&mut self) -> Result<(), OwnableError>;
}
