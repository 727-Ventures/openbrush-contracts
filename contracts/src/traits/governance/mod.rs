// Copyright (c) 2023 Brushfam
// Copyright (c) 2012-2022 Supercolony
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the"Software"),
// to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

use crate::traits::errors::GovernanceError;
use ink::prelude::vec::Vec;
use openbrush::traits::{
    AccountId,
    Balance,
    Timestamp,
};

pub mod governor;

pub use governor::*;

pub mod extensions {
    pub mod governor_counting;
    pub mod governor_quorum;
    pub mod governor_settings;
    pub mod timelock_controller;
}

pub mod utils {
    pub mod votes;
}
pub type ProposalId = [u8; 32];
pub type HashType = [u8; 32];
pub type Selector = [u8; 4];

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, scale::Encode, scale::Decode)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub struct Transaction {
    pub callee: AccountId,
    pub selector: [u8; 4],
    pub input: Vec<u8>,
    pub transferred_value: Balance,
    pub gas_limit: u64,
}

#[derive(scale::Decode, scale::Encode, Default, Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub enum ExecutionStatus {
    #[default]
    NotExecuted,
    Executed,
}

#[derive(scale::Decode, scale::Encode, Default, Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub enum CancelationStatus {
    #[default]
    NotCanceled,
    Canceled,
}

#[derive(scale::Decode, scale::Encode, Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub struct ProposalCore {
    pub proposer: AccountId,
    pub vote_start: Timestamp,
    pub vote_duration: Timestamp,
    pub executed: ExecutionStatus,
    pub canceled: CancelationStatus,
}

impl ProposalCore {
    pub fn new(proposer: AccountId, vote_start: Timestamp, vote_duration: Timestamp) -> Self {
        Self {
            proposer,
            vote_start,
            vote_duration,
            executed: ExecutionStatus::NotExecuted,
            canceled: CancelationStatus::NotCanceled,
        }
    }

    pub fn is_executed(&self) -> bool {
        self.executed == ExecutionStatus::Executed
    }

    pub fn is_canceled(&self) -> bool {
        self.canceled == CancelationStatus::Canceled
    }

    pub fn deadline(&self) -> Result<u64, GovernanceError> {
        let start = self.vote_start.clone();
        let duration = self.vote_duration.clone();

        start.checked_add(duration).ok_or(GovernanceError::DeadlineOverflow)
    }

    pub fn hash(&self) -> [u8; 32] {
        use ink::env::hash;

        let bytes: Vec<u8> = scale::Encode::encode(&self);

        let mut output = <hash::Blake2x256 as hash::HashOutput>::Type::default();
        ink::env::hash_bytes::<hash::Blake2x256>(&bytes[..], &mut output);

        output
    }
}

#[derive(scale::Decode, scale::Encode, Default, Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub enum ProposalState {
    #[default]
    Pending = 1 << 0,
    Active = 1 << 1,
    Canceled = 1 << 2,
    Defeated = 1 << 3,
    Succeeded = 1 << 4,
    Queued = 1 << 5,
    Expired = 1 << 6,
    Executed = 1 << 7,
}

impl ProposalState {
    pub fn u128(self) -> u128 {
        self as u128
    }
}

pub const ALL_PROPOSAL_STATES: u128 = 0b11111111;

#[derive(scale::Decode, scale::Encode, Default, Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub struct ProposalVote {
    pub against_votes: Balance,
    pub for_votes: Balance,
    pub abstain_votes: Balance,
}

#[derive(scale::Decode, scale::Encode, Default, Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout))]
pub enum VoteType {
    #[default]
    Against,
    For,
    Abstain,
}
