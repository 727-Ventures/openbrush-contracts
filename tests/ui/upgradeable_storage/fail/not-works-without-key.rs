use openbrush::traits::AccountId;
pub const STORAGE_KEY: u32 = openbrush::storage_unique_key!(OwnableData);

#[derive(Debug)]
#[openbrush::upgradeable_storage()]
pub struct OwnableData {
    pub owner: AccountId,
    pub _reserved: Option<()>,
}

const PROXY_KEY : u32 = openbrush::storage_unique_key!(ProxyData);

#[derive(Debug)]
#[openbrush::upgradeable_storage()]
pub struct ProxyData {
    pub forward: AccountId,
    pub _reserved: Option<()>,
}

#[derive(Debug)]
#[openbrush::upgradeable_storage()]
pub struct SomeData {
    pub _reserved: Option<()>,
}


fn main() {}