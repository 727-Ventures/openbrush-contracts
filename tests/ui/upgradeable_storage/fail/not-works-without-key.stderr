error: expected expression, found `;`
 --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:5:1
  |
5 | #[openbrush::upgradeable_storage()]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | expected expression
  | while parsing this item list starting here
  | the item list ends here
  |
  = note: this error originates in the attribute macro `openbrush::upgradeable_storage` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected expression, found `;`
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:14:1
   |
14 | #[openbrush::upgradeable_storage()]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | expected expression
   | while parsing this item list starting here
   | the item list ends here
   |
   = note: this error originates in the attribute macro `openbrush::upgradeable_storage` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected expression, found `;`
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:21:1
   |
21 | #[openbrush::upgradeable_storage()]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | expected expression
   | while parsing this item list starting here
   | the item list ends here
   |
   = note: this error originates in the attribute macro `openbrush::upgradeable_storage` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
 --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:7:16
  |
7 |     pub owner: AccountId,
  |                ^^^^^^^^^ expected at least 1 generic argument
  |
note: struct defined here, with at least 1 generic parameter: `KEY`
 --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
  |
  | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
  |            ^^^^^^^^^ --------------
help: add missing generic argument
   |
7  |     pub owner: AKEYccountId,
   |                 +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
 --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:8:20
  |
8 |     pub _reserved: Option<()>,
  |                    ^^^^^^^^^^ expected at least 1 generic argument
  |
note: struct defined here, with at least 1 generic parameter: `KEY`
 --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
  |
  | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
  |            ^^^^^^^^^ --------------
help: add missing generic argument
   |
8  |     pub _reserved: OKEYption<()>,
   |                     +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:16:18
   |
16 |     pub forward: AccountId,
   |                  ^^^^^^^^^ expected at least 1 generic argument
   |
note: struct defined here, with at least 1 generic parameter: `KEY`
  --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
   |
   | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
   |            ^^^^^^^^^ --------------
help: add missing generic argument
   |
16 |     pub forward: AKEYccountId,
   |                   +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:17:20
   |
17 |     pub _reserved: Option<()>,
   |                    ^^^^^^^^^^ expected at least 1 generic argument
   |
note: struct defined here, with at least 1 generic parameter: `KEY`
  --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
   |
   | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
   |            ^^^^^^^^^ --------------
help: add missing generic argument
   |
17 |     pub _reserved: OKEYption<()>,
   |                     +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:23:20
   |
23 |     pub _reserved: Option<()>,
   |                    ^^^^^^^^^^ expected at least 1 generic argument
   |
note: struct defined here, with at least 1 generic parameter: `KEY`
  --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
   |
   | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
   |            ^^^^^^^^^ --------------
help: add missing generic argument
   |
23 |     pub _reserved: OKEYption<()>,
   |                     +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
 --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:5:1
  |
5 | #[openbrush::upgradeable_storage()]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected at least 1 generic argument
  |
note: struct defined here, with at least 1 generic parameter: `KEY`
 --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
  |
  | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
  |            ^^^^^^^^^ --------------
  = note: this error originates in the attribute macro `openbrush::upgradeable_storage` (in Nightly builds, run with -Z macro-backtrace for more info)
help: add missing generic argument
   |
5  | #KEY[openbrush::upgradeable_storage()]
   |  +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:14:1
   |
14 | #[openbrush::upgradeable_storage()]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected at least 1 generic argument
   |
note: struct defined here, with at least 1 generic parameter: `KEY`
  --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
   |
   | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
   |            ^^^^^^^^^ --------------
   = note: this error originates in the attribute macro `openbrush::upgradeable_storage` (in Nightly builds, run with -Z macro-backtrace for more info)
help: add missing generic argument
   |
14 | #KEY[openbrush::upgradeable_storage()]
   |  +++

error[E0107]: this struct takes at least 1 generic argument but 0 generic arguments were supplied
  --> tests/ui/upgradeable_storage/fail/not-works-without-key.rs:21:1
   |
21 | #[openbrush::upgradeable_storage()]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected at least 1 generic argument
   |
note: struct defined here, with at least 1 generic parameter: `KEY`
  --> $CARGO/ink_storage_traits-4.1.0/src/impls/mod.rs
   |
   | pub struct ManualKey<const KEY: Key, ParentKey: StorageKey = ()>(
   |            ^^^^^^^^^ --------------
   = note: this error originates in the attribute macro `openbrush::upgradeable_storage` (in Nightly builds, run with -Z macro-backtrace for more info)
help: add missing generic argument
   |
21 | #KEY[openbrush::upgradeable_storage()]
   |  +++
